### PR TITLE
listener: create internal listener

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
Implements [internal listener](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener.proto#L93).
This PR allows creating server connection but the abilitity to connect to such a listener is located in #18105. 

Additional Description:
Risk Level: LOW. Not impacting existing TCP/UDP listener.

Testing: Mostly unit test since the dispatcher is not yet ready to create pair of internal connections.

Docs Changes:
Release Notes:
Platform Specific Features:
Runtime guard: `envoy.reloadable_features.internal_address` to allow/reject internal listener config.

[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by lambdai** [comment]
Talked to @yanavlasov about this PR.
/assign yanavlasov

File: test/server/BUILD
Comment (yanavlasov) [inline test/server/BUILD:109]: Hmm, seems like huge change to dep list. Is this correct?

File: test/server/BUILD
Comment (yanavlasov) [inline test/server/BUILD:163]: Does this test need all these dependencies?

File: test/server/connection_handler_test.cc
Comment (yanavlasov) [inline test/server/connection_handler_test.cc:7]: Are you sure all these headers are needed?

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Review by yanavlasov — DISMISSED** [review]

**Comment by lambdai** [comment]
Thank you @yanavlasov !

File: source/server/listener_manager_impl.cc
Comment (rojkov) [inline source/server/listener_manager_impl.cc:338]: Probably now we could say that
```suggestion
        fmt::format("listener {} has envoy internal address {}. This runtime feature is disabled",
```

File: test/server/connection_handler_test.cc
Comment (rojkov) [inline test/server/connection_handler_test.cc:1589]: Better use gtest's assertions here and in other places
```suggestion
  ASSERT_TRUE(internal_listener_cb.has_value());
```

**Review by rojkov — COMMENTED** [review]
Thank you! Looks very good. I added some nitpicks and a couple of questions to clarify.

File: source/server/active_internal_listener.cc
Comment (lambdai) [reply source/server/active_internal_listener.cc:41]: Yes, it depends on `close(Network::ConnectionCloseType::NoFlush)`  moving the connection to the dispatcher deferred removal list

https://github.com/envoyproxy/envoy/blob/365acdbe02ed7bfd2f8fdc616d057b430da2621e/source/server/active_stream_listener_base.cc#L119

File: test/server/connection_handler_test.cc
Comment (lambdai) [reply test/server/connection_handler_test.cc:1589]: TIL. Many thanks!

File: examples/tls-inspector/envoy.yaml
Comment (rojkov) [inline examples/tls-inspector/envoy.yaml:33]: This change looks like an accident. Is it related to the PR?

Comment (lambdai) [reply examples/tls-inspector/envoy.yaml:33]: ah, it's unrelated. updating

**Review by rojkov — APPROVED** [review]
Thank you! LGTM

**Comment by lambdai** [comment]
Thank you @yanavlasov @rojkov !
What's the next step?

**Comment by rojkov** [comment]
/assign-from @envoyproxy/senior-maintainers 
for a final pass and merge

File: test/integration/internal_listener_integration_test.cc
Comment (ggreenway) [inline test/integration/internal_listener_integration_test.cc:26]: I think this is the same as `BaseIntegrationTest(GetParam().version, ConfigHelper::tcpProxyConfig())`

File: test/integration/internal_listener_integration_test.cc
Comment (ggreenway) [inline test/integration/internal_listener_integration_test.cc:42]: Why is this needed? Shouldn't it be enough to set the address type to internal?

File: test/integration/internal_listener_integration_test.cc
Comment (ggreenway) [inline test/integration/internal_listener_integration_test.cc:48]: Maybe rename test to `BasicConfigUpdate`?

File: envoy/network/listener.h
Comment (ggreenway) [inline envoy/network/listener.h:113]: It looks like this isn't used yet for anything. Delete until it's needed?

File: source/common/network/listen_socket_impl.h
Comment (ggreenway) [inline source/common/network/listen_socket_impl.h:171]: Why is this false?

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (ggreenway) [inline source/common/tcp_proxy/tcp_proxy.cc:704]: Trace level?

File: test/server/active_internal_listener_test.cc
Comment (ggreenway) [inline test/server/active_internal_listener_test.cc:201]: Is there a notion of draining an internal listener?

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall

/wait

File: test/integration/internal_listener_integration_test.cc
Comment (lambdai) [reply test/integration/internal_listener_integration_test.cc:42]: That's a good point. The internal listener field was supposed to declare some settings, e.g. loop prevention check, also an indicator that Envoy would ignore the tcp socket options. I tend to agree with you that this internal options should be optional

What do you think the major drawback here?

File: source/common/network/listen_socket_impl.h
Comment (lambdai) [reply source/common/network/listen_socket_impl.h:171]: The close does nothing. So I assume either blindly close or close-while-isOpen works as before. I don't see what would check a listen socket isopen.

A flag is_open can be added later if we do need to maintain the isOpen state, but I think we are not there yet.

File: test/server/active_internal_listener_test.cc
Comment (lambdai) [reply test/server/active_internal_listener_test.cc:201]: Sorry, I don't fully get your question.
Our usage of the internal listener includes the cases each internal listener(or filter chain within) represents a L7 service. Since the l7 service could be http, it's nice to close the connection to internal listener only after a certain timeout.

File: envoy/network/listener.h
Comment (lambdai) [reply envoy/network/listener.h:113]: This is used by the ConnectionHandler to determine if the listener type is Internal.
I feel this is cleaner than find the listen address and query the address type then dispatch by the address type.

In fact I almost make the ConnectionHandler not relying on the address type.

**Comment by lambdai** [comment]
@ggreenway I resolve most of them except introducing the empty InternalListenerConfig interface. PTAL

File: test/integration/internal_listener_integration_test.cc
Comment (ggreenway) [reply test/integration/internal_listener_integration_test.cc]: If the user needs to specify a field in the address (setting the type to internal), it should not be required to set this message; that would mean the user has to essentially set the same logical setting in two places. If there are config settings needed for the internal listener, I think it makes sense to put them here. But if there aren't, and the address is set to internal, that should work.

Comment (lambdai) [reply test/integration/internal_listener_integration_test.cc]: I made the change in https://github.com/envoyproxy/envoy/pull/18104/commits/3f6407b5bdd5eed9d9ab94653668329da701b20d

Basically, this protobuf internal listener config is not required. The listen address. BTW:  I also add the test case that this field is present, the listen address must be internal type.

File: envoy/network/listener.h
Comment (ggreenway) [inline envoy/network/listener.h:461]: I don't think this is ever used; delete.

File: envoy/network/listener.h
Comment (ggreenway) [inline envoy/network/listener.h:449]: I think this should be named `InternalListener`. `InternalListenerCallbacks` would be an interface passed to an `InternalListener` so that it can access other parts of the system.

File: source/server/connection_handler_impl.cc
Comment (ggreenway) [inline source/server/connection_handler_impl.cc:164]: assert that the address type is internal?

File: source/common/network/listen_socket_impl.h
Comment (ggreenway) [inline source/common/network/listen_socket_impl.h:157]: Do you need to override bind()?

**Review by ggreenway — CHANGES_REQUESTED** [review]
I think you need to validate listener config to make sure all the tcp-socket specific options aren't set. Such as tcp-fast-open, socket options, keepalive, mptcp, reuse_port, etc.

File: test/server/active_internal_listener_test.cc
Comment (yanavlasov) [reply test/server/active_internal_listener_test.cc:201]: Conceptually drain is only applicable to downstream listeners that accept connections from external clients. Internal listener accepts connections from upstream connections only. As a result I do not think drain would be applicable.

Comment (lambdai) [reply test/server/active_internal_listener_test.cc:201]: Chat with @yanavlasov  offline. Drain is derived from internal listener update and it's good to gracefully close the connections as in this test case.

File: source/server/connection_handler_impl.cc
Comment (lambdai) [reply source/server/connection_handler_impl.cc:165]: done

File: source/server/listener_impl.cc
Comment (ggreenway) [inline source/server/listener_impl.cc:470]: Does anything go wrong if this is thrown after `buildFilterChains` has already run? I can't remember if that has any interaction with the previous config on the listener. If there's any risk, you could move this into validateConfig() to do it earlier.

File: source/server/listener_impl.cc
Comment (ggreenway) [inline source/server/listener_impl.cc:473]: `|| config_.enable_mptcp()`. And maybe `has_api_listener()`? 

For all of the BoolValue's, I think it's still valid if it is present, but set to false. 

This validation is going to be a big mess. Maybe it's not worth doing. Any guidance from @envoyproxy/api-shepherds  on what validation to do for mutually-incompatible options?

**Review by ggreenway — COMMENTED** [review]
/wait-any

File: source/server/listener_impl.cc
Comment (lambdai) [reply source/server/listener_impl.cc:470]: As long as the exception was thrown prior to ListenerImpl::initialize(), there is no risk to wrongly deem the exception as a signal of "listener is wamed up".

We don't cross the line here

File: source/server/listener_impl.cc
Comment (lambdai) [reply source/server/listener_impl.cc:473]: re: it's still valid if it is present. Right, Let me fix immediately.

Regarding the high level validation, I remember we opened an issue to move the tcp only options to a dedicated tcp listener config, though there was no solid move. That would be the long term fix. 
Another fact is that Udp listener doesn't validate tcp options, just ignore.

I will add the quick fix on the validation here and leave api shephereds to the final call.

File: source/server/listener_impl.cc
Comment (lambdai) [reply source/server/listener_impl.cc:470]: api listener is handled early: Api listener config doesn't instantiate ListenerImpl but HttpApiListener
I add a test in listenerimpl pretending that some how a listenerImpl is constructed with Apilistener. Though in production it will never hit :)

Comment (lambdai) [reply source/server/listener_impl.cc:470]: A separate question to @junr03: should we reject api listener with internal address? 
 https://github.com/envoyproxy/envoy/blob/main/source/server/api_listener_impl.h#L39

**Review by yanavlasov — APPROVED** [review]

**Review by ggreenway — APPROVED** [review]
I think this is pretty good. It's possible that we'll want to change some of the details as you complete the other parts of this (large) feature, but this is good enough for moving on to the next steps.

**Comment by lambdai** [comment]
Thank you @ggreenway @yanavlasov

**Pull request merged**
